include/azure_credentials_provider.h

```diff
@@ -19,34 +11,44 @@
 #include <sstream>
 #include <memory>
 
-#  ifdef LINK_AZURE_LIBS
-#    include <azure/core/credentials/credentials.hpp>
-#  endif
+#include <azure/core/datetime.hpp>
+
+#ifdef LINK_AZURE_LIBS
+#include <azure/core/credentials/credentials.hpp>
+#endif
+
+#include "err_constants.h"
+#include "trace_logger.h"
 
 namespace reinforcement_learning
 {
-
 template<typename T>
 class azure_credentials_provider
 {
 public:
   template<typename... Args>
   azure_credentials_provider(Args&&... args) :
-    _creds(std::make_unique<T>(std::forward<Args>(args)...)) {}
+    _creds(std::make_unique<T>(std::forward<Args>(args)...))
+  {
+  }
 
   int get_credentials(const std::vector<std::string>& scopes, std::string& token_out,
-      std::chrono::system_clock::time_point& expiry_out)
+    std::chrono::system_clock::time_point& expiry_out, i_trace* trace)
   {
+    using namespace Azure::Core;
+    using namespace Azure::Core::Credentials;
     try
     {
-      Azure::Core::Credentials::TokenRequestContext request_context;
-      request_context.Scopes = scopes;
+      TokenRequestContext request_context;
+      Context context;
 
-      Azure::Core::Context context;
-      std::cout << "fetching token for " << scopes[0] << std::endl;
-
-      std::lock_guard<std::mutex> lock(_creds_mtx);
-      auto auth = _creds->GetToken(request_context, context);
+      request_context.Scopes = scopes;
+      AccessToken auth;
+      {
+        std::lock_guard<std::mutex> lock(_creds_mtx);
+        auth = _creds->GetToken(request_context, context);
+        TRACE_DEBUG(trace, "azure_credentials_provider: successfully retrieved token");
```

Comment:
move this outside the scope guard since it can potentially call a cout type thing internally?

Reply:
Never mind.  I did not see the later commit

Reply:
yeah, got that; it's pushed please refreshed